lib/net/packets.js

```diff
@@ -957,6 +969,669 @@ class NotFoundPacket extends InvPacket {
+ * @property {Number} startHeight
+ * @property {Hash} stopHash
+ */
+
+class GetCFHeadersPacket extends Packet {
+  /**
+   * Create a `getcfheaders` packet.
+   * @constructor
+   * @param {Number} filterType - Filter type.
+   * @param {Number} startHeight - Start block height.
+   * @param {Hash} stopHash - Stop block hash.
+   */
+
+  constructor(filterType, startHeight, stopHash) {
+    super();
+
+    this.cmd = 'getcfheaders';
+    this.type = exports.types.GETCFHEADERS;
+
+    this.filterType = filterType;
+    this.startHeight = startHeight || null;
+    this.stopHash = stopHash || null;
+  }
+
+  /**
+   * Get serialization size.
+   * @returns {Number}
+   */
+
+  getSize() {
+    return 37;
+  }
+
+  /**
+   * Serialize getcfheaders packet to writer.
+   * @param {BufferWriter} bw
+   */
+
+  toWriter(bw) {
+    bw.writeU8(this.filterType);
+    bw.writeU32(this.startHeight);
+    bw.writeHash(this.stopHash);
+
+    return bw;
+  }
+
+  /**
+   * Serialize getcfheaders packet.
+   * @returns {Buffer}
+   */
+
+  toRaw() {
+    const size = this.getSize();
+    return this.toWriter(bio.write(size)).render();
+  }
+
+  /**
+   * Inject properties from buffer reader.
+   * @private
+   * @param {BufferReader} br
+   * @returns {GetCFHeadersPacket}
+   */
+
+  fromReader(br) {
+    this.filterType = br.readU8();
+    this.startHeight = br.readU32();
+    this.stopHash = br.readHash('hex');
+
+    return this;
+  }
+
+  /**
+   * Inject properties from serialized data.
+   * @private
+   * @param {Buffer} data
+   * @returns {GetCFHeadersPacket}
+   */
+
+  fromRaw(data) {
+    return this.fromReader(bio.read(data));
+  }
+
+  /**
+   * Instantiate getcfheaders packet from buffer reader.
+   * @param {BufferReader} br
+   * @returns {GetCFHeadersPacket}
+   */
+
+  static fromReader(br) {
+    return new this().fromReader(br);
+  }
+
+  /**
+   * Instantiate getcfheaders packet from serialized data.
+   * @param {Buffer} data
+   * @returns {GetCFHeadersPacket}
+   */
+
+  static fromRaw(data) {
+    return new this().fromRaw(data);
+  }
+}
+
+/**
+ * CFHeaders Packet
+ * @extends Packet
+ * @property {Number} filterType
+ * @property {Hash?} stopHash
+ * @property {Hash} previousFilterHeader
+ * @property {Number} filterHashesLength
+ * @property {(Hash[])?} filterHashes
+ */
+
+class CFHeadersPacket extends Packet {
+  /**
+   * Create a `cfheaders` packet.
+   * @constructor
+   * @param {Number} filterType
+   * @param {Hash?} stopHash
+   * @param {Hash} previousFilterHeader
+   * @param {(Hash[])?} filterHashes
+   */
+
+  constructor(filterType, stopHash, previousFilterHeader, filterHashes) {
+    super();
+
+    this.cmd = 'cfheaders';
+    this.type = exports.types.CFHEADERS;
+
+    this.filterType = filterType;
+    this.stopHash = stopHash || null;
+    this.previousFilterHeader = previousFilterHeader || null;
+    this.filterHashes = filterHashes || [];
+    this.filterHashesLength = this.filterHashes.length;
```

Comment:
If we move the check to the top and then if we create a new object with no params and initialize it using raw data using `fromReader` function (as done in the static methods) the `filterHashes` will not be an array and then `filterHashes.length` will be undefined thus the checks should be only be after initializing `this.filterhashes` and in the `fromReader` function

Reply:
Ok that makes sense but now I'm thinking, do we need the length check in the constructor at all? It makes sense in `fromReader()` since that is basically validating input from the wild p2p network. But the other packets with arrays we don't check those rules in the constructor we enforce them in the caller (wherever we construct the packet from data in the first place).

Reply:
I added the check in the constructor in case we create a new packet and if by any chance we add more than the permitted filterHashes it should give an error there only rather than passing it to the peers.